### ISSUE TITLE
Handling push event fails

### ISSUE THREAD

**Issue by amsdard** [SUSE/Portus#1051]
My setup:

haproxy in front to terminate ssl
registry 2.3.1, bound to domain.io, with HTTP_HOST set to https://domain.io but listening on 80 port inside container 
portus 2.0.5  bound to hub.domain.io, fqdn=hub.domain.io

Registry works, I can login, push pull images. Notification is accepted by portus, but for unknown reason when registry asks to authorize request, he tries to make request to himself on port 443 which is not defined by me in any place.

``` bash
Handling Push event:
31.08.2016 13:16:48{
31.08.2016 13:16:48  "id": "b3dc7d4d-466f-4570-bccb-361fb7a6f6bd",
31.08.2016 13:16:48  "timestamp": "2016-08-31T11:16:48.195504737Z",
31.08.2016 13:16:48  "action": "push",
31.08.2016 13:16:48  "target": {
31.08.2016 13:16:48    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
31.08.2016 13:16:48    "size": 8895,
31.08.2016 13:16:48    "digest": "sha256:bab33e13eea1cd4361546180835153fddb14dec4f86d76f2577d34ac0ed59b21",
31.08.2016 13:16:48    "length": 8895,
31.08.2016 13:16:48    "repository": "wiki",
31.08.2016 13:16:48    "url": "https://domain.io/v2/wiki/manifests/sha256:bab33e13eea1cd4361546180835153fddb14dec4f86d76f2577d34ac0ed59b21"
31.08.2016 13:16:48  },
31.08.2016 13:16:48  "request": {
31.08.2016 13:16:48    "id": "f2a6a73b-a69e-450c-8a6d-2c14dc64713e",
31.08.2016 13:16:48    "addr": "10.42.13.72",
31.08.2016 13:16:48    "host": "domain.io",
31.08.2016 13:16:48    "method": "PUT",
31.08.2016 13:16:48    "useragent": "docker/1.12.0 go/go1.6.3 git-commit/8eab29e kernel/4.4.16-boot2docker os/linux arch/amd64 UpstreamClient(Docker-Client/1.11.1 \\(darwin\\))"
31.08.2016 13:16:48  },
31.08.2016 13:16:48  "actor": {
31.08.2016 13:16:48    "name": "koszi"
31.08.2016 13:16:48  },
31.08.2016 13:16:48  "source": {
31.08.2016 13:16:48    "addr": "domain.io:80",
31.08.2016 13:16:48    "instanceID": "ff9dffc2-f38b-4fd2-9edc-329ab3ba59c7"
31.08.2016 13:16:48  }
31.08.2016 13:16:48}
31.08.2016 13:16:48Could not fetch the tag for target {"mediaType"=>"application/vnd.docker.distribution.manifest.v2+json", "size"=>8895, "digest"=>"sha256:bab33e13eea1cd4361546180835153fddb14dec4f86d76f2577d34ac0ed59b21", "length"=>8895, "repository"=>"wiki", "url"=>"https://domain.io/v2/wiki/manifests/sha256:bab33e13eea1cd4361546180835153fddb14dec4f86d76f2577d34ac0ed59b21"}
31.08.2016 13:16:48Reason: Connection refused - connect(2) for "hub.domain.io" port 443
31.08.2016 13:16:48Completed 202 Accepted in 17ms (ActiveRecord: 1.9ms)
31.08.2016 13:16:48[61] 10.42.203.20 - - [31/Aug/2016:11:16:48 +0000] "POST /v2/webhooks/events HTTP/1.1" 202 - 0.0187
```

Some issues:
**"addr": "domain.io:80",** - this from what I've managed to discover is registry bug, as registry indeed listens on 80 inside container but is available at just domain.io. Is this value used by portus in any way?

**31.08.2016 13:16:48Reason: Connection refused - connect(2) for "hub.domain.io" port 443** - why on earth portus makes request to himself on wrong port? It is configured without ssl, so where this port 443 comes from? And if he wants to use 443 so badly why does not use it with https but with plain http?

One last thing, I have removed port from registry hostname as I do not want my images to by called domain.io:5000/wiki nor domain.io:80/wiki but just domain.io/wiki. 

Any idea how to tell portus to make request on correct port and/or scheme? I'm struggling third day with this and no luck.

**Comment by amsdard** [comment]
The only workaround I've found so far is to run socat, which will forward 443 to haproxy. But I really wonder why portus makes request on 443?

**Comment by mssola** [comment]
```
Could not fetch the tag for target {"mediaType"=>"application/vnd.docker.distribution.manifest.v2+json", "size"=>8895, "digest"=>"sha256:bab33e13eea1cd4361546180835153fddb14dec4f86d76f2577d34ac0ed59b21", "length"=>8895, "repository"=>"wiki", "url"=>"https://domain.io/v2/wiki/manifests/sha256:bab33e13eea1cd4361546180835153fddb14dec4f86d76f2577d34ac0ed59b21"}
```

This is done when processing the event notification, in which Portus (depending on the version of the registry) will try to fetch some extra information about the arriving notification. In this case, since you are using 2.3.1, the tag itself was not included in the notification, so Portus makes an API call to the registry in order to fetch the tag itself. In order to do that, it uses the registry configuration that you added when you accessed Portus for the first time (see [this](http://port.us.org/docs/Configuring-the-registry.html)).

So, long story short, Portus is trying to access the registry in the  same way it was given to it on the initial configuration. Notice that on the `hostname` field, the port matters. If you don't set a port on the hostname, 443 will be assumed because most surely you checked the `Use SSL` checkbox when you configured it. In `master` and in the future `2.1` you will be able to fix the registry with a proper edit page. Unfortunately, since you are on 2.0.5, you will have to update the registry by accessing the DB directly.

**Comment by amsdard** [comment]
As far as I know portus notifications will not work with registry higher than 2.3.1 according to https://github.com/SUSE/Portus/issues/905 thus I'm using what works.

In my configuration, portus was talking to registry over plain http (I have linked registry in portus container) thus **Use ssl** was not checked. Anyway why registry configuration would have any effect on how portus is trying to reach itself? 

For me it looks like when notification happens, portus makes request to registry, registry ask for authorization token and then portus makes request to himself to get token. This makes some sense but I wonder where does it gets this 443 port.

I have changed configuration so portus talks to registry over haproxy with ssl, I've checked ssl checkbox so it talks with registry with no errors, but it still tries to reach hub.domain.io:443 which is portus itself so it results with connection refused as portus is listening on port 3000. 

My registry config:
```bash
    REGISTRY_AUTH: token
    REGISTRY_AUTH_TOKEN_ISSUER: hub.domain.io
    REGISTRY_AUTH_TOKEN_REALM: https://hub.domain.io/v2/token
    REGISTRY_AUTH_TOKEN_ROOTCERTBUNDLE: /opt/certs/domain.io/domain.io.pem
    REGISTRY_AUTH_TOKEN_SERVICE: domain.io
    REGISTRY_LOG_LEVEL: debug
    REGISTRY_STORAGE_DELETE_ENABLED: 'true'
    REGISTRY_HTTP_SECRET: secret
    REGISTRY_HTTP_HOST: https://domain.io
    REGISTRY_HTTP_HEADERS_X-CONTENT-TYPE-OPTIONS: '[nosniff]'
    REGISTRY_LOG_FIELDS_SERVICE: registry
    REGISTRY_HEALTH_STORAGEDRIVER_ENABLED: 'false'
    REGISTRY_NOTIFICATIONS_ENDPOINTS: |-
      - name: portus
        url: http://hub.domain.io:3000/v2/webhooks/events
        timeout: 500
        threshold: 5
        backoff: 1
    REGISTRY_STORAGE: azure
    REGISTRY_STORAGE_AZURE_ACCOUNTNAME: accountname
    REGISTRY_STORAGE_AZURE_ACCOUNTKEY: key
    REGISTRY_STORAGE_AZURE_CONTAINER: registry
    REGISTRY_HTTP_ADDR: 0.0.0.0:80
```
Note that portus is linked in registry container under hub.domain.io alias so it can be accessed over plain http.

I'm using sshipway/portus:2.0.5 image with this config:
```bash
    PORTUS_CHECK_SSL_USAGE_ENABLED: 'false'
    PORTUS_GRAVATAR_ENABLED: 'true'
    PORTUS_KEY_PATH: /opt/certs/domain.io/domain.io.key
    PORTUS_MACHINE_FQDN: hub.domain.io
    PORTUS_PASSWORD: pass
    PORTUS_PORT: '3000'
    PORTUS_PRODUCTION_DATABASE: portus
    PORTUS_PRODUCTION_HOST: db
    PORTUS_PRODUCTION_PASSWORD: pass
    PORTUS_PRODUCTION_USERNAME: root
    PORTUS_SECRET_KEY_BASE: secret
    PORTUS_SMTP_ENABLED: 'false'
    REGISTRY_HOSTNAME: domain.io
    REGISTRY_NAME: Registry
    REGISTRY_SSL_ENABLED: 'true'
```

**Comment by EugenMayer** [comment]
@amsdard be aware, the portus image you are using is very old. This release is about 4 months old, so @mssola will deal with issues which might be no longer present.

Nevertheless in #905 i use the most recent master build of portus and it still does not work - so this issue is still present.

I start to grasp that the real issue might be the configuration of users. Your catalog and mine share the same setup idea and maybe thats the reason. @mssola notice, both catalogs have the same roots, so there might be something in the way of the portus setup

**Comment by amsdard** [comment]
@EugenMayer thanks for update. What image are you using then?

**Comment by EugenMayer** [comment]
@amsdard i am using this image https://github.com/EugenMayer/docker-image-portus or https://hub.docker.com/r/eugenmayer/portus/ with this catalog in rancher https://github.com/EugenMayer/docker-rancher-extra-catalogs

**Comment by amsdard** [comment]
@EugenMayer Thanks, but why I can't find it in rancher catalog? I'm new to rancher and I do not grasp everything yet, and hint how to use your catalog in rancher?

**Comment by amsdard** [comment]
@EugenMayer ok, I've found in rancher settings, I will test your catalog setup.

**Comment by amsdard** [comment]
I've tested portus 2.4 from @EugenMayer  repo and this problem with fetching tags does not appear any more, so I guess this was caused by outdated version.

Big thanks for quick help and guidance!

**Comment by EugenMayer** [comment]
@amsdard do you get those event-issues describe in #905 or you rather dont? could you check this?

**Comment by mssola** [comment]
Closing this then.
